### PR TITLE
Problem: wal write is heavy in block replaying

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
    - introduce `store.async-wal` option to write wal asynchronously
    - rename the `memiavl` flag to `store.memiavl`.

In the block replaying test, it reduce the sync time from 2m to 1m10s, 40% reduction.

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: memiavl/db.go [hunk 1]
```diff
@@ -142,29 +168,44 @@ func (db *DB) ApplyUpgrades(upgrades []*TreeNameUpgrade) error {
...
+		db.checkBackgroundSnapshotRewrite(),
+	)
+}
+
+// checkAsyncWAL check the quit signal of async wal writing
+func (db *DB) checkAsyncWAL() error {
+	select {
+	case err := <-db.walQuit:
+		// async wal writing failed, we need to abort the state machine
+		return fmt.Errorf("async wal writing goroutine quit unexpectedly: %w", err)
```
Comment (mmsqe) [inline memiavl/db.go:184]: need handle nil here
```suggestion
		if err != nil {
			// async WAL writing failed, we need to abort the state machine
			return fmt.Errorf("async WAL writing goroutine quit unexpectedly: %w", err)
		}
```

**Review by mmsqe — COMMENTED** [review]
so async is faster than batch?

File: memiavl/db.go (hunk 1 above)
Comment (yihuang) [reply memiavl/db.go:184]: it also treat `nil` as an error here, the async wal goroutine just not expected to quit.

**Comment by yihuang** [comment]
> so async is faster than batch?

yes, batch don't improve much, async improve a lot.

File: memiavl/db.go (hunk 1 above)
Comment (mmsqe) [reply memiavl/db.go:184]: close will return right?

Comment (yihuang) [reply memiavl/db.go:184]: yes, but whenever it quit, the state machine should be aborted

**Pull request merged**